File: code/java/Exercises.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Stream;

public final class Exercises {

    private Exercises() {
    }

    // ---------------------------------------------------------------------
    // firstThenLowerCase
    // ---------------------------------------------------------------------
    public static Optional<String> firstThenLowerCase(
            List<String> strings,
            Predicate<String> predicate
    ) {
        return strings.stream()
                .filter(predicate)
                .findFirst()
                .map(String::toLowerCase);
    }

    // ---------------------------------------------------------------------
    // say(...) -> Say object with .and(...) and .phrase()
    // ---------------------------------------------------------------------
    public static final class Say {
        private final List<String> words;

        private Say(List<String> words) {
            this.words = List.copyOf(words);
        }

        public Say and(String word) {
            List<String> next = new ArrayList<>(words.size() + 1);
            next.addAll(words);
            next.add(word);
            return new Say(next);
        }

        public String phrase() {
            return String.join(" ", words);
        }
    }

    public static Say say() {
        return new Say(List.of());
    }

    public static Say say(String firstWord) {
        return new Say(List.of(firstWord));
    }

    // ---------------------------------------------------------------------
    // meaningfulLineCount
    // ---------------------------------------------------------------------
    public static int meaningfulLineCount(String filename) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            Stream<String> lines = reader.lines();
            long count = lines.filter(line -> !isCommentLine(line)).count();
            return (int) count;
        } catch (FileNotFoundException e) {
            // Let FileNotFoundException propagate so the test sees it
            throw e;
        }
    }

    // A "comment line" is one whose first non-whitespace char is '#'
    private static boolean isCommentLine(String line) {
        int i = 0;
        while (i < line.length() && Character.isWhitespace(line.charAt(i))) {
            i++;
        }
        if (i < line.length() && line.charAt(i) == '#') {
            return true;
        }
        return false;
    }
}

// =====================================================================
// Quaternion (record, immutable, with ZERO/I/J/K, plus/times/conjugate/toString)
// =====================================================================

record Quaternion(double a, double b, double c, double d) {

    static final Quaternion ZERO = new Quaternion(0.0, 0.0, 0.0, 0.0);
    static final Quaternion I    = new Quaternion(0.0, 1.0, 0.0, 0.0);
    static final Quaternion J    = new Quaternion(0.0, 0.0, 1.0, 0.0);
    static final Quaternion K    = new Quaternion(0.0, 0.0, 0.0, 1.0);

    Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    Quaternion plus(Quaternion other) {
        return new Quaternion(
                this.a + other.a,
                this.b + other.b,
                this.c + other.c,
                this.d + other.d
        );
    }

    Quaternion times(Quaternion other) {
        double a1 = this.a;
        double b1 = this.b;
        double c1 = this.c;
        double d1 = this.d;
        double a2 = other.a;
        double b2 = other.b;
        double c2 = other.c;
        double d2 = other.d;

        return new Quaternion(
                a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
                a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
                a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
                a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2
        );
    }

    private static boolean nonZero(double x) {
        return Math.abs(x) > 0.0;
    }

    @Override
    public String toString() {
        double a = this.a;
        double b = this.b;
        double c = this.c;
        double d = this.d;

        if (!nonZero(a) && !nonZero(b) && !nonZero(c) && !nonZero(d)) {
            return "0";
        }

        StringBuilder sb = new StringBuilder();

        if (nonZero(a)) {
            sb.append(Double.toString(a));
        }

        appendTerm(sb, b, "i");
        appendTerm(sb, c, "j");
        appendTerm(sb, d, "k");

        return sb.toString();
    }

    private static void appendTerm(StringBuilder sb, double coeff, String symbol) {
        if (!nonZero(coeff)) {
            return;
        }
        boolean first = sb.length() == 0;
        double mag = Math.abs(coeff);
        String magStr = (mag == 1.0) ? "" : Double.toString(mag);

        if (first) {
            if (coeff < 0) {
                sb.append("-").append(magStr).append(symbol);
            } else {
                sb.append(magStr).append(symbol);
            }
        } else {
            String sign = coeff < 0 ? "-" : "+";
            sb.append(sign).append(magStr).append(symbol);
        }
    }
}

// =====================================================================
// BinarySearchTree (sealed interface + Empty + Node)
// =====================================================================

sealed interface BinarySearchTree permits Empty, Node {
    int size();
    boolean contains(String value);
    BinarySearchTree insert(String value);
}

final class Empty implements BinarySearchTree {

    Empty() {
    }

    @Override
    public int size() {
        return 0;
    }

    @Override
    public boolean contains(String value) {
        return false;
    }

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(new Empty(), value, new Empty());
    }

    @Override
    public String toString() {
        return "()";
    }
}

final class Node implements BinarySearchTree {

    private final BinarySearchTree left;
    private final String value;
    private final BinarySearchTree right;

    Node(BinarySearchTree left, String value, BinarySearchTree right) {
        this.left = left;
        this.value = value;
        this.right = right;
    }

    @Override
    public int size() {
        return left.size() + 1 + right.size();
    }

    @Override
    public boolean contains(String v) {
        int cmp = v.compareTo(value);
        if (cmp == 0) {
            return true;
        } else if (cmp < 0) {
            return left.contains(v);
        } else {
            return right.contains(v);
        }
    }

    @Override
    public BinarySearchTree insert(String v) {
        int cmp = v.compareTo(value);
        if (cmp == 0) {
            return this;
        } else if (cmp < 0) {
            return new Node(left.insert(v), value, right);
        } else {
            return new Node(left, value, right.insert(v));
        }
    }

    @Override
    public String toString() {
        String leftStr = (left instanceof Empty) ? "" : left.toString();
        String rightStr = (right instanceof Empty) ? "" : right.toString();
        return "(" + leftStr + value + rightStr + ")";
    }
}

File: code/kotlin/Exercises.kt
import java.io.File
import java.io.IOException
import kotlin.math.abs

fun firstThenLowerCase(strings: List<String>, satisfying: (String) -> Boolean): String? =
    strings.firstOrNull(satisfying)?.lowercase()

data class Say(private val words: List<String>) {
    fun and(word: String): Say = Say(words + word)
    val phrase: String
        get() = words.joinToString(" ")
}

fun say(): Say = Say(emptyList())
fun say(word: String): Say = Say(listOf(word))

@Throws(IOException::class)
fun meaningfulLineCount(filename: String): Long {
    val file = File(filename)
    file.bufferedReader().use { reader ->
        return reader
            .lineSequence()
            .filter { !isCommentLine(it) }
            .count()
            .toLong()
    }
}

private fun isCommentLine(line: String): Boolean {
    for (ch in line) {
        if (!ch.isWhitespace()) {
            return ch == '#'
        }
    }
    return false
}

data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {

    init {
        require(!a.isNaN() && !b.isNaN() && !c.isNaN() && !d.isNaN()) {
            "Coefficients cannot be NaN"
        }
    }

    fun coefficients(): List<Double> = listOf(a, b, c, d)

    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)

    operator fun plus(other: Quaternion): Quaternion =
        Quaternion(
            a + other.a,
            b + other.b,
            c + other.c,
            d + other.d
        )

    operator fun times(other: Quaternion): Quaternion {
        val a1 = a
        val b1 = b
        val c1 = c
        val d1 = d
        val a2 = other.a
        val b2 = other.b
        val c2 = other.c
        val d2 = other.d

        return Quaternion(
            a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
            a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
            a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
            a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2
        )
    }

    override fun toString(): String {
        fun nonZero(x: Double) = abs(x) > 0.0

        if (!nonZero(a) && !nonZero(b) && !nonZero(c) && !nonZero(d)) {
            return "0"
        }

        val sb = StringBuilder()

        if (nonZero(a)) {
            sb.append(a.toString())
        }

        fun appendTerm(coeff: Double, symbol: String) {
            if (!nonZero(coeff)) return
            val first = sb.isEmpty()
            val mag = abs(coeff)
            val magStr = if (mag == 1.0) "" else mag.toString()
            if (first) {
                if (coeff < 0) {
                    sb.append("-").append(magStr).append(symbol)
                } else {
                    sb.append(magStr).append(symbol)
                }
            } else {
                val sign = if (coeff < 0) "-" else "+"
                sb.append(sign).append(magStr).append(symbol)
            }
        }

        appendTerm(b, "i")
        appendTerm(c, "j")
        appendTerm(d, "k")

        return sb.toString()
    }

    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }
}

sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(v: String): Boolean
    fun insert(v: String): BinarySearchTree

    object Empty : BinarySearchTree {
        override fun size(): Int = 0

        override fun contains(v: String): Boolean = false

        override fun insert(v: String): BinarySearchTree =
            Node(Empty, v, Empty)

        override fun toString(): String = "()"
    }

    data class Node(
        val left: BinarySearchTree,
        val value: String,
        val right: BinarySearchTree
    ) : BinarySearchTree {

        override fun size(): Int = left.size() + 1 + right.size()

        override fun contains(v: String): Boolean {
            val cmp = v.compareTo(value)
            return when {
                cmp == 0 -> true
                cmp < 0  -> left.contains(v)
                else     -> right.contains(v)
            }
        }

        override fun insert(v: String): BinarySearchTree {
            val cmp = v.compareTo(value)
            return when {
                cmp == 0 -> this
                cmp < 0  -> Node(left.insert(v), value, right)
                else     -> Node(left, value, right.insert(v))
            }
        }

        override fun toString(): String {
            val leftStr = if (left === Empty) "" else left.toString()
            val rightStr = if (right === Empty) "" else right.toString()
            return "($leftStr$value$rightStr)"
        }
    }
}
